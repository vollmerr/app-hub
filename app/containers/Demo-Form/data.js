import {
  FieldText,
  FieldSelect,
  FieldDate,
  FieldRadios,
  FieldChecks,
  FieldFile,
} from 'components/Form';

import { normalizeNumber } from './normalize';

/**
 * Example of how a const would typically be used.
 * Using a const allows the following:
 *
 * 1. Easily change field name (such as if API changes it we must also)
 * We can just change the string in this one place instead of multiple.
 * (ex from 'textDefault' to 'defaultText')
 *
 * 2. Consistent names and intellisense support. This const is used 3
 * times in the example bellow. Having to manually type all of them is
 * error prone and does not allow auto completion.
 */
const SECTION_TEXT = 'sectionText';
const FIELD_TEXT_DEFAULT = 'textDefault';

export default {
  /**
   * Order by section name to allow easily looping over in the same order and lookup
   * const from above used for consistency (in the first value).
   * Would nomrally be placed under bySection, but above for readability of example
   */
  allSections: [SECTION_TEXT, 'select', 'dates', 'radios', 'checks', 'file'],
  // allSections: ['checks'],
  /**
   * All sections of form (unordered)
   */
  bySection: {
    /**
     * Section of form (in this case all the text fields)
     * example using the const from above, notice [] around it
     */
    [SECTION_TEXT]: {
      /**
       * Title that will be used in the heading above section
       */
      title: 'Text',
      /**
       * Field data such that the name is the key
       */
      byName: {
        [FIELD_TEXT_DEFAULT]: { // example using the const from above, notice [] around it
          label: 'Default Text',
          name: FIELD_TEXT_DEFAULT, // example using the const above for name
          placeholder: 'Enter text...',
          component: FieldText,
        },
        textRequried: {
          required: true,
          label: 'Requried Text',
          name: 'textRequried',
          placeholder: 'Enter text...',
          component: FieldText,
        },
        textArea: {
          multiline: true,
          label: 'Multiline Text',
          name: 'textArea',
          placeholder: 'Enter text...',
          component: FieldText,
        },
        textDisabled: {
          required: true, // required for testing purposes (should render as not required due to being disabled)
          disabled: true,
          label: 'Disabled Text',
          name: 'textDisabled',
          placeholder: 'Enter text...',
          component: FieldText,
        },
        textNormalized: {
          label: 'Normalized Numbers (####-###)',
          name: 'textNormalized',
          placeholder: 'Enter numbers...',
          normalize: normalizeNumber,
          component: FieldText,
        },
      },
      /**
       * Order by name to allow easily looping over in the same order and lookup
       * Again, const from above used for consistency (in the first value).
       */
      allNames: [FIELD_TEXT_DEFAULT, 'textRequried', 'textArea', 'textDisabled', 'textNormalized'],
    },
    select: {
      title: 'Select / Dropdown',
      byName: {
        selectDefault: {
          label: 'Default Select',
          name: 'selectDefault',
          placeholder: 'Enter text...',
          component: FieldSelect,
          options: [
            { key: 'N/A', text: 'Does Not Apply' },
            { key: 'Under', text: 'Under' },
            { key: 'Over', text: 'Over' },
          ],
        },
        selectRequired: {
          required: true,
          label: 'Requried Select',
          name: 'selectRequired',
          placeholder: 'Enter text...',
          component: FieldSelect,
          options: [
            { key: 'N/A', text: 'Does Not Apply' },
            { key: 'Under', text: 'Under' },
            { key: 'Over', text: 'Over' },
          ],
        },
        selectDisabled: {
          required: true, // required for testing purposes (should render as not required due to being disabled)
          disabled: true,
          label: 'Disabled Select',
          name: 'selectDisabled',
          placeholder: 'Enter text...',
          component: FieldSelect,
          options: [
            { key: 'N/A', text: 'Does Not Apply' },
            { key: 'Under', text: 'Under' },
            { key: 'Over', text: 'Over' },
          ],
        },
      },
      allNames: ['selectDefault', 'selectRequired', 'selectDisabled'],
    },
    radios: {
      title: 'Radio Buttons',
      byName: {
        radiosDefault: {
          label: 'Default Radios',
          name: 'radiosDefault',
          component: FieldRadios,
          options: [
            { key: 'N/A', text: 'Does Not Apply' },
            { key: 'Under', text: 'Under' },
            { key: 'Over', text: 'Over' },
          ],
        },
        radiosRequired: {
          required: true,
          label: 'Required Radios',
          name: 'radiosRequired',
          component: FieldRadios,
          options: [
            { key: 'N/A', text: 'Does Not Apply' },
            { key: 'Under', text: 'Under' },
            { key: 'Over', text: 'Over' },
          ],
        },
        radiosDisabled: {
          required: true, // required for testing purposes (should render as not required due to being disabled)
          disabled: true,
          label: 'Disabled Radios',
          name: 'radiosDisabled',
          component: FieldRadios,
          options: [
            { key: 'N/A', text: 'Does Not Apply' },
            { key: 'Under', text: 'Under' },
            { key: 'Over', text: 'Over' },
          ],
        },
      },
      allNames: ['radiosDefault', 'radiosRequired', 'radiosDisabled'],
    },
    dates: {
      title: 'Date Picker',
      byName: {
        dateDefault: {
          label: 'Default Date',
          name: 'dateDefault',
          placeholder: 'Select Date...',
          component: FieldDate,
        },
        dateRequired: {
          required: true,
          label: 'Required Date',
          name: 'dateRequired',
          placeholder: 'Select Date...',
          component: FieldDate,
        },
        dateDisabled: {
          required: true, // required for testing purposes (should render as not required due to being disabled)
          disabled: true,
          label: 'Disabled Date',
          name: 'dateDisabled',
          placeholder: 'Select Date...',
          component: FieldDate,
        },
      },
      allNames: ['dateDefault', 'dateRequired', 'dateDisabled'],
    },
    checks: {
      title: 'Check Boxes',
      byName: {
        checkDefault: {
          label: 'Default Check Boxes',
          name: 'checkDefault',
          component: FieldChecks,
          options: [
            { key: 'N/A', text: 'Does Not Apply' },
            { key: 'Under', text: 'Under' },
            { key: 'Over', text: 'Over' },
          ],
        },
        checkRequired: {
          required: true,
          label: 'Required Check Boxes',
          name: 'checkRequired',
          component: FieldChecks,
          options: [
            { key: 'N/A', text: 'Does Not Apply' },
            { key: 'Under', text: 'Under' },
            { key: 'Over', text: 'Over' },
          ],
        },
        checkDisabled: {
          required: true, // required for testing purposes (should render as not required due to being disabled)
          disabled: true,
          label: 'Disabled Check Boxes',
          name: 'checkDisabled',
          component: FieldChecks,
          options: [
            { key: 'N/A', text: 'Does Not Apply' },
            { key: 'Under', text: 'Under' },
            { key: 'Over', text: 'Over' },
          ],
        },
      },
      allNames: ['checkDefault', 'checkRequired', 'checkDisabled'],
    },
    file: {
      title: 'File Picker',
      byName: {
        fileDefault: {
          label: 'Default File Picker',
          name: 'fileDefault',
          placeholder: 'Select File...',
          component: FieldFile,
        },
        fileRequired: {
          required: true,
          label: 'Required File Picker',
          name: 'fileRequired',
          placeholder: 'Select File...',
          component: FieldFile,
        },
        fileDisabled: {
          required: true, // required for testing purposes (should render as not required due to being disabled)
          disabled: true,
          label: 'Disabled File Picker',
          name: 'fileDisabled',
          placeholder: 'Select File...',
          component: FieldFile,
        },
      },
      allNames: ['fileDefault', 'fileRequired', 'fileDisabled'],
    },
  },
};
