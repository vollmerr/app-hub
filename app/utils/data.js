/**
 * Formats an item object to the correct format
 * for rendering
 *
 * @param {array} item      - item to format
 * @param {string} name     - internal field name for lookup
 * @param {object} field    - field format info
 * @param {object} enums    - display names for enums
 *
 * @return {string}         - formatted item
 */
export function formatItem(item, name, field, enums = {}) {
  let content = item[name];
  // convert all data to array for mapping over
  if (!Array.isArray(content)) {
    content = [content];
  }
  // go through each item, mapping to the correct format
  content = content.map((x) => {
    // is enum mapping
    if (enums[name]) {
      return enums[name][x];
    }
    // is date
    if (field.data && field.data.type === 'DATE') {
      return isNaN(Date.parse(x)) ? '' : new Date(x).toISOString().substr(0, 10);
    }
    return x;
  });
  // join them back as string seperated by commas
  return content.join(', ');
}

/**
 * Formats an array of item objects to the correct format
 * for rendering
 *
 * @param {array} items     - items to format
 * @param {object} fields   - lookup table of fields
 * @param {object} enums    - display names for enums
 *
 * @return {array}          - formatted items as objects of strings
 */
export function formatItems(items, fields, enums) {
  const formattedItems = [];

  items.forEach((item, i) => {
    formattedItems[i] = {};
    Object.keys(item).forEach((k) => {
      formattedItems[i][k] = formatItem(item, k, fields[k], enums);
    });
  });

  return formattedItems;
}

/**
 * Maps a object to the format Office-UI-Fabric-React DetailsList`s expect
 *
 * @param {object} obj    - object to map
 * @param {array} include - keys to include, if empty includes all
 * @param {array} exclude - keys to exclude
 *
 * @return {object}       - mapped object to use in List
 */
export function mapToColumns(obj, include = [], exclude = []) {
  const keys = include.length ?
    include :
    Object.keys(obj);

  return keys
    .filter((key) => !exclude.includes(key))
    .map((key) => ({
      key: obj[key].name,
      name: obj[key].label,
      fieldName: obj[key].name,
      isResizable: true,
      ariaLabel: obj[key].ariaLabel,
      minWidth: obj[key].minWidth || 100,
      maxWidth: obj[key].maxWidth || 200,
      data: obj[key].data || {},
    }));
}
