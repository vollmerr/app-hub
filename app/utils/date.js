export const dateWithoutTime = (date) => {
  if (date && !isNaN(new Date(date))) {
    const d = new Date(date);
    d.setHours(0, 0, 0, 0);
    return d;
  }
  return undefined;
};
