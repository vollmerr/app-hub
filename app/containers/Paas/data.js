import { types, mapToColumns } from 'utils/data';

import { AUTH, APPS, APP_LIST, REPORT, BUTTONS } from './constants';


export const authorization = {
  [AUTH.ID]: {
    label: 'ID',
    name: AUTH.ID,
    ariaLabel: 'ID of the authorization',
  },
  [AUTH.SID]: {
    label: 'SID',
    name: AUTH.SID,
    ariaLabel: 'SID of the Employee',
  },
  [AUTH.FULL_NAME]: {
    label: 'Full Name',
    name: AUTH.FULL_NAME,
    ariaLabel: 'Full Name of the Employee',
    minWidth: 100,
    maxWidth: 150,
    isSortedDescending: true,
  },
  [AUTH.EMAIL]: {
    label: 'Email',
    name: AUTH.EMAIL,
    ariaLabel: 'Date the acknowledgment takes effect',
  },
  [AUTH.POS_NUMBER]: {
    label: 'Position Number',
    name: AUTH.POS_NUMBER,
    ariaLabel: 'Position Number of the Employee',
  },
  [AUTH.MANAGER_SID]: {
    label: 'Manager SID',
    name: AUTH.MANAGER_SID,
    ariaLabel: 'AD SID of the Employee\'s Manager',
  },
  [AUTH.MANAGER_NAME]: {
    label: 'Manager Name',
    name: AUTH.MANAGER_NAME,
    ariaLabel: 'Full Name of the Employee\'s Manager',
  },
  [AUTH.STATUS]: {
    label: 'Status',
    name: AUTH.STATUS,
    ariaLabel: 'Status of the Authorization',
  },
  [AUTH.LAST_MODIFIED]: {
    label: 'Last Modified',
    name: AUTH.LAST_MODIFIED,
    ariaLabel: 'Date the Authorization was last modified',
    data: {
      type: types.date,
    },
  },
  [AUTH.LAST_APPROVED]: {
    label: 'Date Authorized',
    name: AUTH.LAST_APPROVED,
    ariaLabel: 'Date of last Authorization denial or approval',
    data: {
      type: types.date,
    },
  },
  [AUTH.DATE_CREATED]: {
    label: 'Date Created',
    name: AUTH.DATE_CREATED,
    ariaLabel: 'Date the Authorization was created',
    data: {
      type: types.date,
    },
  },
  [APPS.APP_1]: {
    label: 'App 1 Name',
    name: APPS.APP_1,
    ariaLabel: 'App 1 AriaLabel',
    minWidth: 80,
    maxWidth: 100,
    data: {
      render: 'authorizationToggle',
    },
  },
  [APPS.APP_2]: {
    label: 'App 2 Name',
    name: APPS.APP_2,
    ariaLabel: 'App 2 AriaLabel',
    minWidth: 80,
    maxWidth: 100,
    data: {
      render: 'authorizationToggle',
    },
  },
  [APPS.APP_3]: {
    label: 'App 3 Name',
    name: APPS.APP_3,
    ariaLabel: 'App 3 AriaLabel',
    minWidth: 80,
    maxWidth: 100,
    data: {
      render: 'authorizationToggle',
    },
  },
  [APPS.APP_4]: {
    label: 'App 4 Name',
    name: APPS.APP_4,
    ariaLabel: 'App 4 AriaLabel',
    minWidth: 80,
    maxWidth: 100,
    data: {
      render: 'authorizationToggle',
    },
  },
};


const buttons = {
  [BUTTONS.APPROVE]: {
    name: BUTTONS.APPROVE,
    ariaLabel: 'Authorize all applications for a given user',
    minWidth: 60,
    maxWidth: 60,
    notSortable: true,
    data: {
      render: 'approvalButton',
    },
  },
  [BUTTONS.DENY]: {
    name: BUTTONS.DENY,
    ariaLabel: 'Deny all applications for a given user',
    minWidth: 60,
    maxWidth: 60,
    notSortable: true,
    data: {
      render: 'denyButton',
    },
  },
};

export const currentFieldsApi = [
  AUTH.ID,
  AUTH.SID,
  ...APP_LIST,
];

const currentFields = [
  AUTH.FULL_NAME,
  ...APP_LIST,
  BUTTONS.APPROVE,
  BUTTONS.DENY,
];

const previousFields = [
  AUTH.FULL_NAME,
  ...APP_LIST,
  AUTH.LAST_APPROVED,
];

export const currentColumns = mapToColumns({ ...authorization, ...buttons }, currentFields);
export const previousColumns = mapToColumns(authorization, previousFields);


const reportFields = [
  AUTH.FULL_NAME,
  ...APP_LIST,
  AUTH.MANAGER_NAME,
  AUTH.LAST_APPROVED,
];

export const reportColumns = {
  [REPORT.APPROVED]: mapToColumns(authorization, reportFields),
  [REPORT.DENIED]: mapToColumns(authorization, reportFields),
  [REPORT.PENDING]: mapToColumns(authorization, reportFields),
  [REPORT.NO_MANAGER]: mapToColumns(authorization, reportFields),
};
