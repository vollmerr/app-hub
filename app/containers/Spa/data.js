import { types, mapToColumns } from '../../utils/data';
import { isFutureDate } from '../../utils/validate'; // TODO: joi

import {
  FieldText,
  FieldDate,
  FieldChecks,
  FieldFile,
  FieldRadios,
} from '../../components/Form';

import * as C from './constants';


export const acknowledgment = {
  [C.ACK.ID]: {
    label: 'ID',
    name: C.ACK.ID,
    ariaLabel: 'ID of the acknowledgment',
    minWidth: 20,
    maxWidth: 40,
  },
  [C.ACK.TITLE]: {
    label: 'Title',
    required: true,
    name: C.ACK.TITLE,
    placeholder: 'Enter acknowledgment title',
    ariaLabel: 'Title of the acknowledgment',
    component: FieldText,
  },
  [C.ACK.STATUS]: {
    label: 'Status',
    name: C.ACK.STATUS,
    ariaLabel: 'Status of the acknowledgment',
    maxWidth: 100,
  },
  [C.ACK.START_DATE]: {
    label: 'Start Date',
    required: true,
    name: C.ACK.START_DATE,
    placeholder: 'Select the start date',
    ariaLabel: 'Date the acknowledgment takes effect',
    component: FieldDate,
    validate: [isFutureDate],
    maxWidth: 100,
    data: {
      type: types.date,
    },
  },
  [C.ACK.END_DATE]: {
    label: 'End Date',
    required: true,
    name: C.ACK.END_DATE,
    placeholder: 'Select the end date',
    ariaLabel: 'Date the acknowledgment expires',
    component: FieldDate,
    validate: [isFutureDate],
    maxWidth: 100,
    data: {
      type: types.date,
    },
  },
  [C.ACK.CREATOR_GROUP]: {
    label: 'Creator Group',
    required: true,
    name: C.ACK.CREATOR_GROUP,
    ariaLabel: 'AD group that created the acknowledgment',
    component: FieldRadios,
    options: [],
  },
  [C.ACK.TARGET_GROUPS]: {
    label: 'Target Group(s)',
    required: true,
    name: C.ACK.TARGET_GROUPS,
    ariaLabel: 'AD groups that must acknowledge the policy',
    component: FieldChecks,
    options: [],
  },
  [C.ACK.DETAILS]: {
    label: 'Description',
    required: true,
    multiline: true,
    name: C.ACK.DETAILS,
    placeholder: 'Enter acknowledgment description',
    ariaLabel: 'Description of the acknowledgment',
    component: FieldText,
  },
  [C.ACK.STATEMENT]: {
    label: 'Acknowledgment Statement',
    required: true,
    name: C.ACK.STATEMENT,
    placeholder: 'Enter the acknowledgment statement',
    ariaLabel: 'Statement that will be placed under the details of the acknowledgment',
    component: FieldText,
  },
  [C.ACK.FILE_NAME]: {
    label: 'Attachment Name',
    required: false,
    name: C.ACK.FILE_NAME,
    placeholder: 'Enter attachment name',
    ariaLabel: 'Name of the file that must be read before acknowledging',
    component: FieldText,
  },
  [C.ACK.FILE_CONTENT]: {
    label: 'Attachment',
    required: false,
    name: C.ACK.FILE_CONTENT,
    placeholder: 'Attach a File',
    ariaLabel: 'Attach the file that must be read before acknowledging',
    component: FieldFile,
  },
};


export const recipient = {
  [C.RECIPIENT.ID]: {
    label: 'ID',
    name: C.RECIPIENT.ID,
    ariaLabel: 'ID of the Recipient',
    maxWidth: 100,
  },
  [C.RECIPIENT.SID]: {
    label: 'SID',
    name: C.RECIPIENT.SID,
    ariaLabel: 'SID of the recipient',
    maxWidth: 100,
  },
  [C.RECIPIENT.SAM]: {
    label: 'SAM',
    name: C.RECIPIENT.SAM,
    ariaLabel: 'SAM account of the recipient',
    maxWidth: 100,
  },
  [C.RECIPIENT.FIRST_NAME]: {
    label: 'First Name',
    name: C.RECIPIENT.FIRST_NAME,
    ariaLabel: 'First name of the recipient',
    maxWidth: 100,
  },
  [C.RECIPIENT.LAST_NAME]: {
    label: 'Last Name',
    name: C.RECIPIENT.LAST_NAME,
    ariaLabel: 'Last name of the recipient',
    maxWidth: 100,
  },
  [C.RECIPIENT.EMAIL]: {
    label: 'Email',
    name: C.RECIPIENT.EMAIL,
    ariaLabel: 'Email of the recipient',
  },
  [C.RECIPIENT.MANAGER_SID]: {
    label: 'Manager SID',
    name: C.RECIPIENT.MANAGER_SID,
    ariaLabel: 'SID of the recipients manager',
  },
  [C.RECIPIENT.MANAGER_NAME]: {
    label: 'Manager',
    name: C.RECIPIENT.MANAGER_NAME,
    ariaLabel: 'Full name of the recipients manager',
  },
  [C.RECIPIENT.ACK_ID]: {
    label: 'Acknowledgment ID',
    name: C.RECIPIENT.ACK_DATE,
    ariaLabel: 'ID of the acknowledgment',
  },
  [C.RECIPIENT.ACK_DATE]: {
    label: 'Acknowledgment Date',
    name: C.RECIPIENT.ACK_DATE,
    ariaLabel: 'Date the recipient acknowledged the policy',
    minWidth: 150,
    data: {
      type: types.date,
    },
  },
  [C.RECIPIENT.FIRST_REMINDER_DATE]: {
    label: 'First Reminder Date',
    name: C.RECIPIENT.FIRST_REMINDER_DATE,
    ariaLabel: 'Date of recipients first acknowledgment reminder',
    data: {
      type: types.date,
    },
  },
  [C.RECIPIENT.SECOND_REMINDER_DATE]: {
    label: 'Second Reminder Date',
    name: C.RECIPIENT.SECOND_REMINDER_DATE,
    ariaLabel: 'Date of recipients second acknowledgment reminder',
    data: {
      type: types.date,
    },
  },
  [C.RECIPIENT.FINAL_REMINDER_DATE]: {
    label: 'Final Reminder Date',
    name: C.RECIPIENT.FINAL_REMINDER_DATE,
    ariaLabel: 'Date of recipients final acknowledgment reminder',
    data: {
      type: types.date,
    },
  },
};


// FORM
export const formFields = {
  left: [
    C.ACK.TITLE,
    C.ACK.START_DATE,
    C.ACK.END_DATE,
    C.ACK.CREATOR_GROUP,
    C.ACK.TARGET_GROUPS,
  ],
  right: [
    C.ACK.STATEMENT,
    C.ACK.DETAILS,
    C.ACK.FILE_NAME,
    C.ACK.FILE_CONTENT,
  ],
};


// COLUMNS
const fields = {
  ...recipient,
  ...acknowledgment, // ACK ID WILL OVERRIDE RECIPIENT
};

const homeFields = [
  // C.ACK.ID,
  C.ACK.TITLE,
  C.ACK.STATUS,
  C.ACK.START_DATE,
  C.ACK.END_DATE,
  C.RECIPIENT.ACK_DATE,
  C.ACK.FILE_NAME,
];
const homeExcludes = [
  C.RECIPIENT.ACK_DATE,
];

const adminFields = [
  // C.ACK.ID,
  C.ACK.TITLE,
  C.ACK.STATUS,
  C.ACK.START_DATE,
  C.ACK.END_DATE,
  C.ACK.FILE_NAME,
];

const reportFields = [
  C.RECIPIENT.FIRST_NAME,
  C.RECIPIENT.LAST_NAME,
  C.RECIPIENT.EMAIL,
  C.RECIPIENT.MANAGER_NAME,
  C.RECIPIENT.ACK_DATE,
];
const reportPendingExcludes = [
  C.RECIPIENT.ACK_DATE,
];

export const homeColumns = {
  pending: mapToColumns(fields, homeFields, homeExcludes),
  previous: mapToColumns(fields, homeFields),
};

export const adminColumns = mapToColumns(fields, adminFields);

export const reportColumns = {
  [C.REPORT.PENDING]: mapToColumns(fields, reportFields, reportPendingExcludes),
  [C.REPORT.PREVIOUS]: mapToColumns(fields, reportFields),
};

// CSV FILES
const adminCsvFields = [
  C.RECIPIENT.FIRST_NAME,
  C.RECIPIENT.LAST_NAME,
  C.RECIPIENT.EMAIL,
  C.RECIPIENT.MANAGER_NAME,
  C.RECIPIENT.ACK_DATE,
  C.RECIPIENT.FIRST_REMINDER_DATE,
  C.RECIPIENT.SECOND_REMINDER_DATE,
  C.RECIPIENT.FINAL_REMINDER_DATE,
];

export const adminCsv = {
  fields: adminCsvFields,
  fieldNames: adminCsvFields.map((x) => recipient[x].label),
};
