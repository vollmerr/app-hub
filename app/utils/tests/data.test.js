import { types, formatItem, formatItems, mapToColumns } from '../data';
import { formattedDate } from '../date';


const enums = {
  enum: {
    a: 'test 1',
    b: 'test 1',
  },
};

const fields = {
  date: { data: { type: types.date } },
  enum: {},
  arr: {},
};

const items = [
  { arr: ['a', 'b'], date: null, enum: 'a' },
  { arr: [], date: '2001/01/02', enum: 'b' },
  { arr: ['c'], date: new Date('2017/10/01'), enum: null },
];

describe('formatItem', () => {
  it('should format enums as their enum value', () => {
    items.forEach((item) => {
      expect(formatItem(item, 'enum', fields.date, enums)).toEqual(enums.enum[item.enum] || '');
    });
  });

  it('should format dates', () => {
    items.forEach((item) => {
      expect(formatItem(item, 'date', fields.date)).toEqual(formattedDate(item.date));
    });
  });

  it('should format array data as a comma seperated string', () => {
    items.forEach((item) => {
      expect(formatItem(item, 'arr', {})).toEqual(item.arr.join(', '));
    });
  });
});

describe('formatItems', () => {
  it('should format all items correctly', () => {
    const formattedItems = formatItems(items, fields, enums);

    expect(formattedItems).toMatchSnapshot();
    expect(formattedItems.every((item) => (
      Object.values(item).every((v) => typeof v === 'string')
    ))).toEqual(true);
  });
});


const toMap = {
  a: { name: 'testName1', label: 'testLabel1', otherStuff: 'testing...' },
  b: { name: 'testName2', label: 'testLabel2', otherStuff: 'testing...' },
  c: { name: 'testName3', label: 'testLabel3', otherStuff: 'testing...' },
};

describe('mapToColumns', () => {
  it('should map the object to the correct format for office-ui-fabric-react DetailsList`s', () => {
    const actual = mapToColumns(toMap);
    expect(actual).toMatchSnapshot();
    expect(actual.length).toEqual(3);
  });

  it('should only use keys in `include` if provided', () => {
    const actual = mapToColumns(toMap, ['a', 'c']);
    expect(actual).toMatchSnapshot();
    expect(actual.length).toEqual(2);
  });

  it('should not use keys in `exclude` if provided', () => {
    const actual = mapToColumns(toMap, [], ['b', 'c']);
    expect(actual).toMatchSnapshot();
    expect(actual.length).toEqual(1);
  });
});
