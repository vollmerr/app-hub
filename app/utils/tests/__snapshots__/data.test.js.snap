// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatItems should format all items correctly 1`] = `
Array [
  Object {
    "arr": "a, b",
    "date": "",
    "enum": "test 1",
  },
  Object {
    "arr": "",
    "date": "2001-01-02",
    "enum": "test 1",
  },
  Object {
    "arr": "c",
    "date": "2017-10-01",
    "enum": "",
  },
]
`;

exports[`mapToColumns should map the object to the correct format for office-ui-fabric-react DetailsList\`s 1`] = `
Array [
  Object {
    "ariaLabel": undefined,
    "data": Object {},
    "fieldName": "testName1",
    "isResizable": true,
    "isSortedDescending": undefined,
    "key": "testName1",
    "maxWidth": 200,
    "minWidth": 100,
    "name": "testLabel1",
    "notSortable": false,
  },
  Object {
    "ariaLabel": undefined,
    "data": Object {},
    "fieldName": "testName2",
    "isResizable": true,
    "isSortedDescending": undefined,
    "key": "testName2",
    "maxWidth": 200,
    "minWidth": 100,
    "name": "testLabel2",
    "notSortable": false,
  },
  Object {
    "ariaLabel": undefined,
    "data": Object {},
    "fieldName": "testName3",
    "isResizable": true,
    "isSortedDescending": undefined,
    "key": "testName3",
    "maxWidth": 200,
    "minWidth": 100,
    "name": "testLabel3",
    "notSortable": false,
  },
]
`;

exports[`mapToColumns should not use keys in \`exclude\` if provided 1`] = `
Array [
  Object {
    "ariaLabel": undefined,
    "data": Object {},
    "fieldName": "testName1",
    "isResizable": true,
    "isSortedDescending": undefined,
    "key": "testName1",
    "maxWidth": 200,
    "minWidth": 100,
    "name": "testLabel1",
    "notSortable": false,
  },
]
`;

exports[`mapToColumns should only use keys in \`include\` if provided 1`] = `
Array [
  Object {
    "ariaLabel": undefined,
    "data": Object {},
    "fieldName": "testName1",
    "isResizable": true,
    "isSortedDescending": undefined,
    "key": "testName1",
    "maxWidth": 200,
    "minWidth": 100,
    "name": "testLabel1",
    "notSortable": false,
  },
  Object {
    "ariaLabel": undefined,
    "data": Object {},
    "fieldName": "testName3",
    "isResizable": true,
    "isSortedDescending": undefined,
    "key": "testName3",
    "maxWidth": 200,
    "minWidth": 100,
    "name": "testLabel3",
    "notSortable": false,
  },
]
`;
